File: find_smallest_letter_greater_than_target.rs
// https://leetcode.com/problems/find-smallest-letter-greater-than-target/description/

impl Solution {
    pub fn next_greatest_letter(letters: &mut Vec<char>, target: &mut char) -> char {
        let mut i = 0;
        let mut j = letters.len() as i32 - 1;
        while i <= j {
            let mid = i + (j - i) / 2;
            if letters[mid as usize] <= *target {
                i = mid + 1;
            } else {
                j = mid - 1;
            }
        }

        if i as usize == letters.len() {
            letters[0]
        } else {
            letters[i as usize]
        }
    }
}

File: algorithms/easy/find_smallest_letter_greater_than_target.kt
class Solution {
  fun nextGreatestLetter(letters: CharArray, target: Char): Char {
    var l = 0
    var r = letters.size - 1
    var ans = 0
    while (l <= r) {
      val mid = (l + r) / 2
      if (letters[mid] > target) {
        ans = mid
        r = mid - 1
      } else {
        l = mid + 1
      }
    }

    return letters[ans]
  }
}
